info-compiler/src/main/java/com/google/sps/webcrawler/WebCrawler.java

```diff
@@ -155,80 +160,118 @@ public void compileNewsArticle(String candidateName, String candidateId) throws
     try {
       HttpGet httpGet = new HttpGet(request);
       JsonObject json = InfoCompiler.requestHttpAndBuildJsonResponse(httpClient, httpGet);
-      return extractUrlsFromCustomSearchJson(json);
+      return extractUrlsAndMetadataFromCustomSearchJson(json);
     } catch (IOException e) {
       System.out.println("[ERROR] Error occurred with fetching URLs from Custom Search: " + e);
       return Arrays.asList();
     }
   }
 
-  // @TODO [Extract other metadata and store into the updated NewsArticle structure.]
   /**
    * Parses {@code json}, which is in Google Custom Search's JSON response format, and extracts
-   * news articles' URLs (URLs from the source website, instead of from News.google).
+   * news articles' URLs (URLs from the source website, instead of from News.google) and metadata,
+   * including the publisher and published date. Packages extracted data into {@code NewsArticle}.
+   * Set news articles' priority based on the order in which they are returned by Custom Search.
    */
-  List<URL> extractUrlsFromCustomSearchJson(JsonObject json) {
-    List<URL> urls = new ArrayList<>(CUSTOM_SEARCH_RESULT_COUNT);
+  List<NewsArticle> extractUrlsAndMetadataFromCustomSearchJson(JsonObject json) {
+    List<NewsArticle> newsArticles = new ArrayList<>(CUSTOM_SEARCH_RESULT_COUNT);
     JsonArray searchResults = json.getAsJsonArray("items");
+    int priority = 1;
     for (JsonElement result : searchResults) {
       JsonObject metadata =
           (JsonObject)
               (((JsonObject) result)
                    .getAsJsonObject("pagemap")
                        .getAsJsonArray("metatags")
                            .get(0));
+      String url = metadata.get(CUSTOM_SEARCH_URL_METATAG).getAsString();
+      String publisher = extractPublisherMetadata(metadata);
+      Date publishedDate = extractPublishedDateMetadata(metadata);
+      newsArticles.add(new NewsArticle(url, publisher, publishedDate, priority));
+      priority++;
+    }
+    return newsArticles;
+  }
+
+  /**
+   * Extracts the publisher from {@code metadata} by examining in order
+   * {@code CUSTOM_SEARCH_PUBLISHER_METATAGS}. Returns null if no matching metatags are found.
+   */
+  private String extractPublisherMetadata(JsonObject metadata) {
+    for (String potentialPublisherMetatag : CUSTOM_SEARCH_PUBLISHER_METATAGS) {
       try {
-        urls.add(
-            new URL(
-                metadata.get(CUSTOM_SEARCH_URL_METATAG).getAsString()));
-      } catch (Exception e) {}
+        return metadata.get(potentialPublisherMetatag).getAsString();
+      } catch (NullPointerException e) {}
```

Comment:
Prefer to use `.has()` first instead of just catching the `NullPointerException`.